// angular-cd-timer/bundles/angular-cd-timer.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core')) :
    typeof define === 'function' && define.amd ? define('angular-cd-timer', ['exports', '@angular/core'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global['angular-cd-timer'] = {}, global.ng.core));
}(this, (function (exports, core) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from) {
        for (var i = 0, il = from.length, j = to.length; i < il; i++, j++)
            to[j] = from[i];
        return to;
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var CdTimerComponent = /** @class */ (function () {
        function CdTimerComponent(elt, renderer) {
            this.elt = elt;
            this.renderer = renderer;
            this.onStart = new core.EventEmitter();
            this.onStop = new core.EventEmitter();
            this.onTick = new core.EventEmitter();
            this.onComplete = new core.EventEmitter();
            // Initialization
            this.autoStart = true;
            this.startTime = 0;
            this.endTime = 0;
            this.timeoutId = null;
            this.countdown = false;
            this.format = 'default';
        }
        CdTimerComponent.prototype.ngAfterViewInit = function () {
            var ngContentNode = this.elt.nativeElement.lastChild; // Get last child, defined by user or span
            this.ngContentSchema = ngContentNode ? ngContentNode.nodeValue : '';
            if (this.autoStart === undefined || this.autoStart === true) {
                this.start();
            }
        };
        CdTimerComponent.prototype.ngOnDestroy = function () {
            this.resetTimeout();
        };
        /**
         * Start the timer
         */
        CdTimerComponent.prototype.start = function () {
            this.initVar();
            this.resetTimeout();
            this.computeTimeUnits();
            this.startTickCount();
            this.onStart.emit(this);
        };
        /**
         * Resume the timer
         */
        CdTimerComponent.prototype.resume = function () {
            this.resetTimeout();
            this.startTickCount();
        };
        /**
         * Stop the timer
         */
        CdTimerComponent.prototype.stop = function () {
            this.clear();
            this.onStop.emit(this);
        };
        /**
         * Reset the timer
         */
        CdTimerComponent.prototype.reset = function () {
            this.initVar();
            this.resetTimeout();
            this.clear();
            this.computeTimeUnits();
            this.renderText();
        };
        /**
         * Get the time information
         * @returns TimeInterface
         */
        CdTimerComponent.prototype.get = function () {
            return {
                seconds: this.seconds,
                minutes: this.minutes,
                hours: this.hours,
                days: this.days,
                timer: this.timeoutId,
                tick_count: this.tickCounter
            };
        };
        /**
         * Initialize variable before start
         */
        CdTimerComponent.prototype.initVar = function () {
            this.startTime = this.startTime || 0;
            this.endTime = this.endTime || 0;
            this.countdown = this.countdown || false;
            this.tickCounter = this.startTime;
            // Disable countdown if start time not defined
            if (this.countdown && this.startTime === 0) {
                this.countdown = false;
            }
            // Determine auto format
            if (!this.format) {
                this.format = (this.ngContentSchema.length > 5) ? 'user' : 'default';
            }
        };
        /**
         * Reset timeout
         */
        CdTimerComponent.prototype.resetTimeout = function () {
            if (this.timeoutId) {
                clearInterval(this.timeoutId);
            }
        };
        /**
         * Render the time to DOM
         */
        CdTimerComponent.prototype.renderText = function () {
            var e_1, _a;
            var outputText;
            if (this.format === 'user') {
                // User presentation
                var items = {
                    'seconds': this.seconds,
                    'minutes': this.minutes,
                    'hours': this.hours,
                    'days': this.days
                };
                outputText = this.ngContentSchema;
                try {
                    for (var _b = __values(Object.keys(items)), _c = _b.next(); !_c.done; _c = _b.next()) {
                        var key = _c.value;
                        outputText = outputText.replace('[' + key + ']', items[key].toString());
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
            }
            else if (this.format === 'intelli') {
                // Intelli presentation
                outputText = '';
                if (this.days > 0) {
                    outputText += this.days.toString() + 'day' + ((this.days > 1) ? 's' : '') + ' ';
                }
                if ((this.hours > 0) || (this.days > 0)) {
                    outputText += this.hours.toString() + 'h ';
                }
                if (((this.minutes > 0) || (this.hours > 0)) && (this.days === 0)) {
                    outputText += this.minutes.toString().padStart(2, '0') + 'min ';
                }
                if ((this.hours === 0) && (this.days === 0)) {
                    outputText += this.seconds.toString().padStart(2, '0') + 's';
                }
            }
            else if (this.format === 'hms') {
                // Hms presentation
                outputText = this.hours.toString().padStart(2, '0') + ':';
                outputText += this.minutes.toString().padStart(2, '0') + ':';
                outputText += this.seconds.toString().padStart(2, '0');
            }
            else {
                // Default presentation
                outputText = this.days.toString() + 'd ';
                outputText += this.hours.toString() + 'h ';
                outputText += this.minutes.toString() + 'm ';
                outputText += this.seconds.toString() + 's';
            }
            this.renderer.setProperty(this.elt.nativeElement, 'innerHTML', outputText);
        };
        CdTimerComponent.prototype.clear = function () {
            this.resetTimeout();
            this.timeoutId = null;
        };
        /**
         * Compute each unit (seconds, minutes, hours, days) for further manipulation
         * @protected
         */
        CdTimerComponent.prototype.computeTimeUnits = function () {
            if (!this.maxTimeUnit || this.maxTimeUnit === 'day') {
                this.seconds = Math.floor(this.tickCounter % 60);
                this.minutes = Math.floor((this.tickCounter / 60) % 60);
                this.hours = Math.floor((this.tickCounter / 3600) % 24);
                this.days = Math.floor((this.tickCounter / 3600) / 24);
            }
            else if (this.maxTimeUnit === 'second') {
                this.seconds = this.tickCounter;
                this.minutes = 0;
                this.hours = 0;
                this.days = 0;
            }
            else if (this.maxTimeUnit === 'minute') {
                this.seconds = Math.floor(this.tickCounter % 60);
                this.minutes = Math.floor(this.tickCounter / 60);
                this.hours = 0;
                this.days = 0;
            }
            else if (this.maxTimeUnit === 'hour') {
                this.seconds = Math.floor(this.tickCounter % 60);
                this.minutes = Math.floor((this.tickCounter / 60) % 60);
                this.hours = Math.floor(this.tickCounter / 3600);
                this.days = 0;
            }
            this.renderText();
        };
        /**
         * Start tick count, base of this component
         * @protected
         */
        CdTimerComponent.prototype.startTickCount = function () {
            var that = this;
            that.timeoutId = setInterval(function () {
                var counter;
                if (that.countdown) {
                    // Compute finish counter for countdown
                    counter = that.tickCounter;
                    if (that.startTime > that.endTime) {
                        counter = that.tickCounter - that.endTime - 1;
                    }
                }
                else {
                    // Compute finish counter for timer
                    counter = that.tickCounter - that.startTime;
                    if (that.endTime > that.startTime) {
                        counter = that.endTime - that.tickCounter - 1;
                    }
                }
                that.computeTimeUnits();
                var timer = {
                    seconds: that.seconds,
                    minutes: that.minutes,
                    hours: that.hours,
                    days: that.days,
                    timer: that.timeoutId,
                    tick_count: that.tickCounter
                };
                that.onTick.emit(timer);
                if (counter < 0) {
                    that.stop();
                    that.onComplete.emit(that);
                    return;
                }
                if (that.countdown) {
                    that.tickCounter--;
                }
                else {
                    that.tickCounter++;
                }
            }, 1000); // Each seconds
        };
        return CdTimerComponent;
    }());
    CdTimerComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'cd-timer',
                    template: ' <ng-content></ng-content>'
                },] }
    ];
    CdTimerComponent.ctorParameters = function () { return [
        { type: core.ElementRef },
        { type: core.Renderer2 }
    ]; };
    CdTimerComponent.propDecorators = {
        startTime: [{ type: core.Input }],
        endTime: [{ type: core.Input }],
        countdown: [{ type: core.Input }],
        autoStart: [{ type: core.Input }],
        maxTimeUnit: [{ type: core.Input }],
        format: [{ type: core.Input }],
        onStart: [{ type: core.Output }],
        onStop: [{ type: core.Output }],
        onTick: [{ type: core.Output }],
        onComplete: [{ type: core.Output }]
    };

    var CdTimerModule = /** @class */ (function () {
        function CdTimerModule() {
        }
        return CdTimerModule;
    }());
    CdTimerModule.decorators = [
        { type: core.NgModule, args: [{
                    declarations: [CdTimerComponent],
                    imports: [],
                    exports: [CdTimerComponent]
                },] }
    ];

    /*
     * Public API Surface of angular-cd-timer
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.CdTimerComponent = CdTimerComponent;
    exports.CdTimerModule = CdTimerModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
